cmake/EthCompilerSettings.cmake

```diff
@@ -14,56 +14,41 @@
 #
 # These settings then end up spanning all POSIX platforms (Linux, OS X, BSD, etc)
 
-include(EthCheckCXXCompilerFlag)
+include(CableCompileOptions)
 
 if(NOT EMSCRIPTEN)
-	eth_add_cxx_compiler_flag_if_supported(-fstack-protector-strong have_stack_protector_strong_support)
-	if(NOT have_stack_protector_strong_support)
-		eth_add_cxx_compiler_flag_if_supported(-fstack-protector)
-	endif()
+	cable_add_compile_options(IF_SUPPORTED -fstack-protector -fstack-protector-strong)
 endif()
 
-eth_add_cxx_compiler_flag_if_supported(-Wimplicit-fallthrough)
-
-# Prevent the path of the source directory from ending up in the binary via __FILE__ macros.
-eth_add_cxx_compiler_flag_if_supported("-fmacro-prefix-map=${CMAKE_SOURCE_DIR}=/solidity")
-
-# -Wpessimizing-move warns when a call to std::move would prevent copy elision
-# if the argument was not wrapped in a call.  This happens when moving a local
-# variable in a return statement when the variable is the same type as the
-# return type or using a move to create a new object from a temporary object.
-eth_add_cxx_compiler_flag_if_supported(-Wpessimizing-move)
-
-# -Wredundant-move warns when an implicit move would already be made, so the
-# std::move call is not needed, such as when moving a local variable in a return
-# that is different from the return type.
-eth_add_cxx_compiler_flag_if_supported(-Wredundant-move)
-
 if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang"))
 	# Enables all the warnings about constructions that some users consider questionable,
 	# and that are easy to avoid.  Also enable some extra warning flags that are not
 	# enabled by -Wall.   Finally, treat at warnings-as-errors, which forces developers
 	# to fix warnings as they arise, so they don't accumulate "to be fixed later".
-	add_compile_options(-Wall)
-	add_compile_options(-Wextra)
-	add_compile_options(-Werror)
-	add_compile_options(-pedantic)
-	add_compile_options(-Wmissing-declarations)
-	add_compile_options(-Wno-unknown-pragmas)
-	add_compile_options(-Wimplicit-fallthrough)
-	add_compile_options(-Wsign-conversion)
-	add_compile_options(-Wconversion)
-
-	eth_add_cxx_compiler_flag_if_supported(
-		$<$<COMPILE_LANGUAGE:CXX>:-Wextra-semi>
+	cable_add_compile_options(
+		-Wall
+		-Wextra
+		-Werror
+		-pedantic
+		-Wmissing-declarations
+		-Wno-unknown-pragmas
+		-Wimplicit-fallthrough
+		-Wsign-conversion
+		-Wconversion
+		IF_SUPPORTED
+		-Wextra-semi
+		-Wfinal-dtor-non-final-class
+		-Wnewline-eof
+		-Wsuggest-destructor-override
+		-Wduplicated-cond
+		-Wduplicate-enum
+		-Wlogical-op
+		-Wno-unknown-attributes
+		-Wpessimizing-move
+		-Wredundant-move
+		# Prevent the path of the source directory from ending up in the binary via __FILE__ macros.
+		-fmacro-prefix-map=${CMAKE_SOURCE_DIR}=/solidity
```

Comment:
What happens if ``CMAKE_SOURCE_DIR`` contains spaces or a semicolon?

Reply:
This might at least need to be quoted.

Reply:
Quoted it, but can you try it locally perhaps? The flag does not work on macos.

Reply:
Lol, more than one thing broken in that setting apparently:
```
$ pwd
/home/daniel/weird test; dir/solidity/build
$ cmake ..
CMake Error at CMakeLists.txt:17 (include):
  include could not find requested file:

    EthToolchains


CMake Error at CMakeLists.txt:20 (include):
  include could not find requested file:

    EthPolicy


CMake Error at CMakeLists.txt:21 (eth_policy):
  Unknown CMake command "eth_policy".


-- Configuring incomplete, errors occurred!
```

Reply:
Without the semicolon it works, though - and I would guess that cmake itself might be incompatible with paths containing a semicolon.

Reply:
It ends up passing ``"-fmacro-prefix-map=\"/home/daniel/weird test dir/solidity=/solidity\""`` to gcc... which I guess is fine, since it doesn't break, but let me see if it actually replaces paths.

Reply:
Oh, it actually doesn't. ``-fmacro-prefix-map="/home/daniel/weird test dir/solidity=/solidity"`` works, but the escaped quotes break it.

Reply:
I don't see an easy way to fix this without looking into it in more detail.

Reply:
I mean, we could just say: who in their right mind uses paths with spaces, but it'd be nice if it worked nevertheless...

Reply:
```
$ clang -fmacro-prefix-map=""
clang: error: unknown argument: '-fmacro-prefix-map='
clang: error: no input files
$ clang -v
Apple clang version 11.0.0 (clang-1100.0.33.8)
Target: x86_64-apple-darwin19.6.0
```

Based on https://en.wikipedia.org/wiki/Xcode#Xcode_7.0_-_12.x_(since_Free_On-Device_Development) supposedly I have upstream clang 11.0.0.

Reply:
@chfast any comment on quotes/escaping quotes issue @ekpyron is observing?

Reply:
Quoting in CMake is fucked up. I usually just try different things until it works. But in most cases it will automatically quote a variable content.

Reply:
Yeah, stuff like that in cmake is indeed *quite* fucked up... but I haven't found any combination on the user side that would make this work properly - I haven't tried getting it to work by changing `cable_add_compile_options` myself yet, though.

Reply:
CMake even  has a workaround helper `separate_arguments`. It is probably not worth the fight and better to handle `-fmacro-prefix-map` manually.